src/wp-admin/includes/class-wp-upgrader.php

```diff
@@ -947,6 +1005,122 @@ public static function create_lock( $lock_name, $release_timeout = null ) {
+			return false;
+		}
+
+		/**
+		 * Skip any plugin that has "." as its slug.
+		 * A slug of "." will result in a `$src` value ending in a period.
+		 *
+		 * On Windows, this will cause the 'plugins' folder to be moved,
+		 * and will cause a failure when attempting to call `mkdir()`.
+		 */
+		if ( '.' === $args['slug'] ) {
+			return false;
+		}
+
+		$dest_dir = $wp_filesystem->wp_content_dir() . 'upgrade/temp-backup/';
+		// Create the temp-backup directory if it doesn't exist.
+		if ( (
+				! $wp_filesystem->is_dir( $dest_dir )
+				&& ! $wp_filesystem->mkdir( $dest_dir )
+			) || (
+				! $wp_filesystem->is_dir( $dest_dir . $args['dir'] . '/' )
+				&& ! $wp_filesystem->mkdir( $dest_dir . $args['dir'] . '/' )
+			)
+		) {
+			return new WP_Error( 'fs_temp_backup_mkdir', $this->strings['temp_backup_mkdir_failed'] );
+		}
+
+		$src_dir = $wp_filesystem->find_folder( $args['src'] );
+		$src     = trailingslashit( $src_dir ) . $args['slug'];
+		$dest    = $dest_dir . $args['dir'] . '/' . $args['slug'];
+
+		// Delete the temp-backup directory if it already exists.
+		if ( $wp_filesystem->is_dir( $dest ) ) {
+			$wp_filesystem->delete( $dest, true );
+		}
+
+		// Move to the temp-backup directory.
+		if ( ! move_dir( $src, $dest ) ) {
+			return new WP_Error( 'fs_temp_backup_move', $this->strings['temp_backup_move_failed'] );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Restores the plugin/theme from the temp-backup directory.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+	 *
+	 * @param array $args Array of data for the temp-backup. Must include a slug, the source, and directory.
+	 * @return bool|WP_Error
+	 */
+	public function restore_temp_backup( $args ) {
+		global $wp_filesystem;
+
+		if ( empty( $args['slug'] ) || empty( $args['src'] ) || empty( $args['dir'] ) ) {
+			return false;
+		}
+
+		$src      = $wp_filesystem->wp_content_dir() . 'upgrade/temp-backup/' . $args['dir'] . '/' . $args['slug'];
+		$dest_dir = $wp_filesystem->find_folder( $args['src'] );
+		$dest     = trailingslashit( $dest_dir ) . $args['slug'];
+
+		if ( $wp_filesystem->is_dir( $src ) ) {
+			// Cleanup.
+			if ( $wp_filesystem->is_dir( $dest ) && ! $wp_filesystem->delete( $dest, true ) ) {
+				return new WP_Error( 'fs_temp_backup_delete', $this->strings['temp_backup_restore_failed'] );
+			}
+
+			// Move it.
+			if ( ! move_dir( $src, $dest ) ) {
+				return new WP_Error( 'fs_temp_backup_delete', $this->strings['temp_backup_restore_failed'] );
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Deletes a temp-backup.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+	 *
+	 * @param array $args Array of data for the temp-backup. Must include a slug, the source, and directory.
+	 * @return bool
+	 */
+	public function delete_temp_backup( $args ) {
+		global $wp_filesystem;
+
+		if ( empty( $args['slug'] ) || empty( $args['dir'] ) ) {
+			return false;
+		}
```

Comment:
I don't understand this one. My ignorance.

Reply:
It seems the only place these can be user modified is https://github.com/WordPress/wordpress-develop/blob/e5cb4986f1f42a9161f9978463752009b375a688/src/wp-admin/includes/class-wp-upgrader.php#L713

Pretty sure it doesn't need protection or sanitizing.

Reply:
@peterwilsoncc Would [validate_file()](https://developer.wordpress.org/reference/functions/validate_file/) do the trick?

```php
if ( ! empty( $args['slug'] ) || empty( $args['dir'] ) || 0 < validate_file( $args['slug'] ) || 0 < validate_file( $args['dir'] ) ) {
    return false;
}
```

Reply:
Given that the `$args` are essentially passed via the `upgrade_package_options` filter, do we need to do this check or should we need to do this check?

Reply:
I think it would be good to have the check somewhere, but if there's an earlier place already in Core, then it might be an issue for a separate ticket.

Reply:
`$options['destination']` has always been changeable via that filter and core hasn't never done any sanitization/dir traversal protection on that...and I don't think this is any different.  Now, if there is potentially a security issue here, then I think that should be addressed in a separate ticket (since the potential security issue also applies to `$options['destination']`).

Also, I'm not sure what that protection would be...because file paths **outside** of `ABSPATH` have always been allowed and **must** continue to be allowed.. to support setups such as storing all plugins/themes outside of `ABSPATH` and symlink'ing them into `WP_PLUGIN_DIR`.  I do this locally and hook into `upgrader_package_options` to modify `$options['destination']` to be that dir outside of `ABSPATH` where I store all the plugins/themes on my local machine.

Reply:
@SergeyBiryukov thoughts on this??

Reply:
> It seems the only place these can be user modified is ...[`upgrader_package_options`]

If they can't come from GET or POST parameters then I think it's fine. The upgrade nonce includes the plugin as part of the action name, so changing that fails. 

I've tried adding things like `hook_extra[temp_backup][slug]=..%2F..%2F..%2F..` and `temp_backup[slug]=..%2F..%2F..%2F..`  to the query string without success. 

@SergeyBiryukov A logic check here would still be great, thanks :)

Reply:
> If they can't come from GET or POST parameters then I think it's fine.

I don't think GET or POST parameters are involved here, this appears to only be changeable via the `upgrader_package_options` filter as noted above.